model/build.go

```diff
@@ -149,3 +152,25 @@ func RemoveBuild(buildID string) error {
 
 	return errors.Wrap(db.C(BuildsCollection).RemoveId(buildID), "deleting build record")
 }
+
+// Generates a new build ID based on the hash of builder and buildNum
+func NewBuildId(builder string, buildNum int) (string, error) {
+	hasher := md5.New()
+
+	// This depends on the fact that Go's json implementation sorts json keys
+	// lexicographically for maps, which ensures consistent encoding
+	jsonMap := make(map[string]interface{})
+	jsonMap["builder"] = builder
+	jsonMap["buildNum"] = buildNum
+	hashstring, err := json.Marshal(jsonMap)
+
+	if err != nil {
+		return "", errors.Wrap(err, "generating json to hash for build key")
+	}
```

Comment:
Given the result is going to be hashed, why is it desirable to marshal to JSON first?
Like couldn't we just concatenate the builder and buildnum? Or `fmt.Sprintf("%s-%d", builder, buildNum)`

Reply:
Ugh, I spent a fair bit of time going mentally back and forth on this.

The big reason is collisions and forward compatibility. In order to avoid collisions, the function needs to be injective (for every element in the output, there is only one input that will give that output). I thought about doing it with fmt.Sprintf, and there _shouldn't_ be the possibility of collision since buildNum is guaranteed not to have a hyphen (so you can't have ("asdf-123", 456) colliding with ("asdf", "123-456")), but the advantage of JSON is that it's already an encoding with escaping rules and no collisions already in place (I think we are all convinced that json strings can always be unambiguously decoded to a consistent set of key/value pairs). It also means that if we wanted to add a 3rd, optional value that'd be really easy. If for example we added another string, we don't have to figure out a way to unambiguously encode that in an injective way that doesn't collide with the existing encoding of two elements.

Unfortunately, JSON has its own problems, the biggest of which is concerns is that the encoding is not, strictly speaking, actually a function, primarily due to key ordering not being guaranteed to be consistent. Go does sort map keys when outputting to JSON however, so key ordering shouldn't be an issue. I'm not super attached to doing it this way, but I think it's slightly cleaner and safer?

Probably the "best" option is to specify an unambiguous binary encoding, which would probably look like:
FIELDNUM (big endian 4 bytes) FIELDLEN (big endian 4 bytes) TEXT

But that seemed scarier than just using JSON.

Reply:
One could also do a text based encoding that's not JSON and have it be unambiguous as well, but that requires an unambiguous text grammar. One easy way is finding a separator that never occurs in your data (I thought about using the null byte). Another would be to do something similar with "FIELD_NUM FIELD_LEN\n" followed by exactly FIELD_LEN bytes of text. Another would be to pick a separator and an escape sequence and make sure you escape anything that would be a separator (and also that you escape anything that is the escape sequence followed by a separator).

Reply:
I think my second favorite option was probably the binary encoding.

Reply:
Ah, thank you for the edifying explanation.